### ISSUE TITLE
Applying custom styles to source blocks

### ISSUE THREAD

**Issue by ggrossetie** [asciidoctor/asciidoctor-pdf#1486]
Let's say I want to write a Best Practices guide and I want to use a role on source blocks when it's not recommended. For instance:

```
Never enter passwords in plain text on the CLI:

[source.dont,sh]
----
$ mvn --encrypt-master-password P@ssw0rd

$ mvn --encrypt-password P@ssw0rd
----
```

In this case, I want to apply a custom style to highlight the fact that the reader should *not* do that.
Is that possible?

**Comment by mojavelinux** [comment]
This is essentially the same as #483. There's currently no framework/callback in place to apply overrides from a role to a block. But the challenge is knowing what to override, what the precedence to apply, and which blocks support which properties.

**Comment by Mogztter** [comment]
> But the challenge is knowing what to override, what the precedence to apply, and which blocks support which properties.

Maybe I'm missing something but I think the role should  have a higher precedence right?
Not sure about the other type of blocks but for a listing block I think we can support every properties that are available on a role: https://github.com/asciidoctor/asciidoctor-pdf/blob/master/docs/theming-guide.adoc#role

- background-color
- border-color
- border-offset
- border-radius
- border-width
- font-color
- font-family
- font-size
- font-style
- text-decoration
- text-decoration-color
- text-decoration-width

Do you want to implement this feature once for all type of blocks or can I submit a pull request to implement this feature only on listing blocks?

**Comment by mojavelinux** [comment]
Yes, the role would win. But sometimes the key is figuring out where to merge the properties in time for the code to access them. Because sometimes the code goes straight to the theme and then it would end up undoing a previously overridden value. In other words, how the theme properties are accessed has to change.

**Comment by mojavelinux** [comment]
I would like #483 to be implemented first, because it's the simpler of the changes. And then we can do listing blocks after that...going further if the design allows.

**Comment by mojavelinux** [comment]
I documented one way to accomplish this using an extended converter. See https://github.com/asciidoctor/asciidoctor-pdf/issues/959#issuecomment-1128668223

**Comment by mojavelinux** [comment]
The technique for supporting roles per block type is now covered in the docs. See https://docs.asciidoctor.org/pdf-converter/latest/extend/use-cases/#theme-table-using-custom-role

**Comment by mojavelinux** [comment]
@gemkod please direct questions about usage to the project chat at htttps://chat.asciidoctor.org. Thanks.

**Comment by mojavelinux** [comment]
Custom roles for blocks is going to be something we're just going to handle in the extended converter. We just cannot continue to make this converter more complex. It's already a tremendous pain to maintain and we have to draw a line.

If there is something that you cannot accomplish using an extended converter, then we can address that. Buf if it can be done using one of the methods documented on the Use Cases page, then that's how it needs to be done.
